### ISSUE TITLE
Update only part of the weight in a layer

### ISSUE THREAD

**Issue by flydance0217** [ppwwyyxx/tensorpack#424]
Hello,

I am wondering what will be the best way to perform partial weight update of a given layer? This is somewhat similar to my previous question [#342](https://github.com/ppwwyyxx/tensorpack/issues/342), where after every step I want to look at the gradient, and according to some criteria then decide whether a portion of the weight should be updated or if the gradient needs to be overwritten with a different value or not. 

I looked at the previous issues as well as FAQ, but the 3 methods that you and other users have mentioned seen to work on a layer basis, and does not allow partial weight update in a given layer. 

I think PostProcessOptimizer is what I am looking for, though I am not exactly sure how to use it. Can you kindly provide an example under this context?

To summarize and perhaps it will be easier to understand, this is what I was hoping to achieve at the moment:
In a conv layer, only perform the gradient update on first 2 kernels for 2 steps. For the next 2 steps, only the 3rd and 4th kernels are updated, and so on.
For now, I think can just select which kernel I want to update by multiplying the gradient with an index tensor, where the tensor can be manually overwritten in the callback, but I am not sure what is the best approach/place to perform the gradient manipulation.

Thanks,

**Comment by flydance0217** [comment]
Probably should have given more thoughts before posting. A colleague just pointed out I should able to accomplish what I described using ScaleGradient, but instead of passing in a scalar, pass in a tensor instead. 

Will come back if I figure out how it works.

**Comment by ppwwyyxx** [comment]
This feature you need already break the abstraction of a "layer". It's best to then write the layer yourself.

**Comment by flydance0217** [comment]
Hi ppwwyyxx,

Agree with you on the example I gave above,  but it might still be useful in some cases where one may want to manipulate the gradient on-the-fly based on some user-defined criteria. I was able to modify your ScaleGradient code so that the grad tensor is multiplied with a "criteria" tensor to achieve per weight gradient control. This way, I only need to modify the "criteria" tensor to do what I want.

Thanks for all the help

**Comment by ppwwyyxx** [comment]
GradientProcessor is designed for manipulating the gradient. You can subclass `GradientProcessor` and implement something similar to `ScaleGradient` like you've already done.

**Comment by flydance0217** [comment]
I have a follow up question. I can pass in a "criteria" tensor (built as part of the tf.global_variables) consists of only 0 or 1 to achieve per weight gradient control of any layer. However it seems that the ScaleGradient function only loads the content of this tensor once initially when the graph is built (which makes sense) and assume the content remains static.

As an example, initially the "criteria" tensor is set to 1 everywhere, meaning that every weight can be updated. Then at certain point I want to modify this tensor in my callback function so that say the first two rows (assuming its a 2D tensor) become 0, meaning the corresponding two rows of weight will no longer be updated. 

How would you go about approaching this problem? 

Thanks,

Pierce

**Comment by ppwwyyxx** [comment]
If you need to change it frequently then it makes sense to make that tensor one of your inputs.
If not, you can also create an op which modifies that variable, and run that op once a while, maybe through the `RunOp` callback.

**Comment by ppwwyyxx** [comment]
Feel free to reopen if you still have questions.
